k8s/src/main/scala/io/buoyant/k8s/EventLogging.scala

```diff
@@ -1,55 +1,95 @@
-    if (values.nonEmpty) {
-      log.debug(
-        "k8s ns %s service %s %s %ss",
-        ns, srv, verb, descriptor
-      )
-      values.foreach(logAction(verb, descriptor, format))
-    }
-
-  private[this] def formatMapping[A, B](kv: (A, B)): String = kv match {
+  @inline private[this] def formatMapping[A, B](kv: (A, B)): String = kv match {
     case (name: String, port: Int) => s"'$name' to port $port"
     case (name: String, to) => s"'$name' to $to"
     case (fromPort, to) => s"from $fromPort to $to"
   }
 
-  @inline def deletion[A, B](deleted: Map[A, B]): Unit =
-    logActions("deleted", "port mapping", formatMapping)(deleted)
+  @inline private[this] def defaultFormat[A](a: A): String = a.toString
 
-  @inline def addition[A, B](added: Map[A, B]): Unit =
-    logActions("added", "port mapping", formatMapping)(added)
+  protected def mkDeletion[A](
+    descriptor: String,
+    format: A => String = defaultFormat[A](_)
+  )(
+    deleted: Iterable[A]
+  ): Unit =
+    if (deleted.nonEmpty) {
+      log.debug(
+        "k8s ns %s service %s deleted %ss",
+        ns, srv, descriptor
+      )
+      if (log.isLoggable(Logger.TRACE)) {
+        deleted.foreach(logAction("deleted", descriptor, format))
+      }
+    }
 
-  protected def logModification[A](descriptor: String)(mods: Iterable[(A, A)]): Unit =
-    if (mods.nonEmpty) {
+  def deletion[A, B](deleted: Map[A, B]): Unit =
+    mkDeletion("port mapping", formatMapping)(deleted)
+
+  protected def mkNewState[A](
+    descriptor: String,
+    format: A => String = defaultFormat[A](_)
+  )(
+    old: Set[A],
+    newState: Set[A]
+  ): Unit =
+    if (old.nonEmpty && old != newState) {
       log.debug(
         "k8s ns %s service %s modified %ss",
         ns, srv, descriptor
       )
-      for {
-        (old, replacement) <- mods
-      } logAction[A](s"replaced $old with", descriptor, _.toString)(replacement)
+      if (log.isLoggable(Logger.TRACE)) {
+        // added
+        (newState -- old).foreach {
+          logAction("added", descriptor, format)
+        }
+        // deleted
+        (old -- newState).foreach {
+          logAction("deleted", descriptor, format)
+        }
+
+      }
     }
 
-  def modification[A, B](old: Map[A, B], replacement: Map[A, B]): Unit =
-    if (old.nonEmpty && replacement.nonEmpty) {
+  def newState[A, B](old: Map[A, B], newState: Map[A, B]): Unit =
+    if (old.nonEmpty && old != newState) {
       log.debug(
         "k8s ns %s service %s modified port mappings",
         ns, srv
       )
-      for {
-        key <- old.keySet.intersect(replacement.keySet)
-        oldValue <- old.get(key)
-        newValue <- replacement.get(key)
-        if oldValue != newValue
-      } log.ifTrace(s"k8s ns $ns service $srv remapped port '$key' from $oldValue to $newValue")
+      if (log.isLoggable(Logger.TRACE)) {
+
+        val (remapped: Set[Option[String]], removed: Set[Option[String]]) = (for {
```

Comment:
unzip is cute but this is probably more readable by building up :scream: mutable sets.

Reply:
noooooo, the horror